pkg/validator/validators.go

```diff
@@ -283,3 +283,91 @@ func MultiClusterServiceValidator(req *admissionv1.AdmissionRequest) (*admission
 
 	return nil, nil
 }
+
+// meshRootCertificateValidator validates the MeshRootCertificate CRD.
+func meshRootCertificateValidator(req *admissionv1.AdmissionRequest) (*admissionv1.AdmissionResponse, error) {
+	mrcSetting := &configv1alpha2.MeshRootCertificate{}
+
+	if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(mrcSetting); err != nil {
+		return nil, err
+	}
+
+	//todo(schristoff:)fake data for now
+	allMRCs := []configv1alpha2.MeshRootCertificate{}
+	numAppliedMRC := len(allMRCs)
+
+	switch req.Operation {
+	case admissionv1.Delete:
+		if numAppliedMRC == 1 {
+			return nil, errors.Errorf("must have more than one Mesh Root Certificate to delete")
+		}
+
+		for _, v := range allMRCs {
+			if mrcSetting.Name == v.Name {
+				currentMRC := &v
+				return validateMRCdelete(currentMRC, mrcSetting)
+			}
+		}
+		return nil, errors.Errorf("cannot find mesh root certificate with name %v", mrcSetting.Name)
+
+	case admissionv1.Create:
+		if numAppliedMRC == 2 {
+			return nil, errors.Errorf("cannot create more then two Mesh Root Certificates")
+		}
+
+		return validateMRCcreate(mrcSetting)
+
+	case admissionv1.Update:
+		for _, v := range allMRCs {
+			if mrcSetting.Name == v.Name {
+				currentMRC := &v
+				return validateMRCupdate(currentMRC, mrcSetting)
+			}
+		}
+		return nil, errors.Errorf("cannot find mesh root certificate with name %v", mrcSetting.Name)
+	}
+
+	return nil, nil
+}
+
+func validateMRCdelete(current *configv1alpha2.MeshRootCertificate,
+	applied *configv1alpha2.MeshRootCertificate) (*admissionv1.AdmissionResponse, error) {
+	//todo(schristoff): will we ever be stuck in these states? how/where do we get out?
+	// if osm gets stuck, will it touch this and get blocked?
+	switch current.Status.RotationStage {
+	case constants.MRCStageIssuing:
+		return nil, errors.Errorf("cannot delete certificate %v in stage %v", current.Name, current.Status.RotationStage)
+	case constants.MRCStageValidating:
+		return nil, errors.Errorf("cannot delete certificate %v in stage %v", current.Name, current.Status.RotationStage)
+	default:
+		return nil, nil
+	}
+}
+
+func validateMRCupdate(current *configv1alpha2.MeshRootCertificate,
+	applied *configv1alpha2.MeshRootCertificate) (*admissionv1.AdmissionResponse, error) {
+	//prevent out of order status changes
+	// we can do:
+	// validating -> issuing
+	// issuing -> validating
+	// complete -> none?
+	switch current.Status.RotationStage {
+	//where are the const for these?
+	case constants.MRCStageValidating:
+		if applied.Status.RotationStage == constants.MRCStateComplete {
```

Comment:
Here we are comparing rotation stage and state. We have to compare the rotation stage with other rotation stages like issuing and validating.